src/core/eme/utils/init_data_store.ts

```diff
@@ -14,98 +14,149 @@
+                             initDataHash : number;
+                             initData: Uint8Array;
+                             value : T; }>;
 
+  /** Construct a new InitDataStore.  */
   constructor() {
-    this._namedTypeData = {};
-    this._unnamedTypeData = {};
+    this._storage = [];
+  }
+
+  /**
+   * Returns all stored value, in the order in which they have been stored.
+   * @returns {Array}
+   */
+  public getAll() : T[] {
+    return this._storage.map(item => item.value);
+  }
+
+  /**
+   * Returns the number of stored values.
+   * @returns {number}
+   */
+  public getLength() : number {
+    return this._storage.length;
   }
 
   /**
-   * Returns true if this instance has the given initData stored.
+   * Returns the element associated with the given initData and initDataType.
+   * Returns `undefined` if not found.
    * @param {Uint8Array} initData
    * @param {string|undefined} initDataType
-   * @returns {boolean}
+   * @returns {*}
    */
   public get(
-    initDataType : string|undefined,
-    initData : Uint8Array
-  ) : T | undefined {
-    if (!isNonEmptyString(initDataType)) {
-      return this._unnamedTypeData[hashBuffer(initData)];
-    }
+    initData : Uint8Array,
+    initDataType : string|undefined
+  ) : T|undefined {
+    const initDataHash = hashBuffer(initData);
+    const index = this._findIndex(initData, initDataType, initDataHash);
+    return index >= 0 ? this._storage[index].value :
+                        undefined;
+  }
 
-    const forDataType = this._namedTypeData[initDataType];
-    if (forDataType == null) {
-      return undefined;
+  /**
+   * Add to the store a value linked to the corresponding initData and
+   * initDataType.
+   * If a value was already stored linked to those, replace it.
+   * @param {Uint8Array} initData
+   * @param {string|undefined} initDataType
+   * @returns {boolean}
+   */
+  public store(
+    initData : Uint8Array,
+    initDataType : string | undefined,
+    value : T
+  ) : void {
+    const initDataHash = hashBuffer(initData);
+    const indexOf = this._findIndex(initData, initDataType, initDataHash);
+    if (indexOf >= 0) {
+      // this._storage contains the stored value in the same order they have
+      // been put. So here we want to remove the previous element and re-push
+      // it to the end.
+      this._storage.splice(indexOf, 1);
     }
-    return forDataType[hashBuffer(initData)];
+    this._storage.push({ initData, initDataType, initDataHash, value });
   }
 
   /**
-   * Add initialization data to this memory.
+   * Add to the store a value linked to the corresponding initData and
+   * initDataType.
+   * If a value linked to those was already stored, do nothing and returns
+   * `false`.
+   * If not, add the value and return `true`.
    * @param {Uint8Array} initData
    * @param {string|undefined} initDataType
+   * @returns {boolean}
    */
-  public set(
-    initDataType : string|undefined,
+  public storeIfNone(
```

Comment:
Is it really necessary to have this function ?
Aren't a `get` and `store` sufficient ?

Reply:
This is kind of an over-optimization :p.

This avoid doing the hashing and lookup two times in a row when you just want to know if a data is found, and if not store one (which you do many times with the `InitDataStore` such as when checking which initialization data has already been encountered).

The name may be poor though.

I understand if you say that this might be too complex here.
I may be doing too much Rust on the side where the concept of [HashMap entries](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html) to allow the same micro-optim appealed to me :P).

Reply:
It's not necesseraly too complex, but it's not a practice we are used to in the RxPlayer (to have an alternative 'set' or 'store' for a store/cache, etc. If you want to keep it, no problem.